src/NuGet.Clients/NuGet.SolutionRestoreManager/InfoBarService.cs

```diff
@@ -0,0 +1,165 @@
+            {
+                _closeFromHide = false;
+            }
+        }
+
+        protected async Task<IVsInfoBarHost?> GetInfoBarHostAsync(CancellationToken cancellationToken)
+        {
+            IVsUIShell? uiShell = await _asyncServiceProvider.GetServiceAsync<SVsUIShell, IVsUIShell>(throwOnFailure: false);
+            if (uiShell == null)
+            {
+                return null;
+            }
+
+            // Ensure that we are on the UI thread before interacting with the Solution Explorer UI Element
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            if (ErrorHandler.Failed(uiShell.FindToolWindow((uint)__VSFINDTOOLWIN.FTW_fFindFirst, VSConstants.StandardToolWindows.SolutionExplorer, out var windowFrame)))
+            {
+                return null;
+            }
+
+            object tempObject;
+            if (ErrorHandler.Failed(windowFrame.GetProperty((int)__VSFPROPID7.VSFPROPID_InfoBarHost, out tempObject)))
+            {
+                return null;
+            }
+
+            return tempObject as IVsInfoBarHost;
+        }
+
+        protected InfoBarModel GetInfoBarModel()
+        {
+            return new InfoBarModel(
+                new IVsInfoBarTextSpan[] {
+                    new InfoBarTextSpan(Resources.InfoBar_TextMessage),
+                },
+                new IVsInfoBarActionItem[] {
+                    new InfoBarHyperlink(Resources.InfoBar_HyperlinkMessage),
+                },
+                KnownMonikers.StatusWarning);
+        }
+
+        internal async Task CreateAndShowInfoBarAsync(CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (_closed || _visible)
+            {
+                return;
+            }
+
+            try
+            {
+                IVsInfoBarUIFactory? infoBarFactory = await _asyncServiceProvider.GetServiceAsync<SVsInfoBarUIFactory, IVsInfoBarUIFactory>(throwOnFailure: false);
+                if (infoBarFactory == null)
+                {
+                    var exception = new NullReferenceException(nameof(infoBarFactory));
+                    await TelemetryUtility.PostFaultAsync(exception, typeof(IVsInfoBarUIFactory).Name, nameof(CreateAndShowInfoBarAsync));
+                    return;
+                }
+
+                IVsInfoBarHost? infoBarHost = await GetInfoBarHostAsync(cancellationToken);
+                if (infoBarHost == null)
+                {
+                    var exception = new NullReferenceException(nameof(infoBarHost));
+                    await TelemetryUtility.PostFaultAsync(exception, typeof(IVsInfoBarHost).Name, nameof(CreateAndShowInfoBarAsync));
+                    return;
+                }
+
+                // Ensure that we are on the UI thread before interacting with the UI
+                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
This is bad for performance. `GetInfoBarHostAsync` switches to the UI thread, which means that if it's called from the thread pool, the `await` will schedule the continuation on the thread pool, but then it calls `jtf.SwitchToMainThreadAsync`, so it needs to be scheduled on the UI thread again.

If you switch to the main thread before calling `GetInfoBarHost`, then you reduce latency since this code will not need to ping-pong between the thread pool and UI thread.

Finally, if you switch to the thread pool before calling `GetInfoBarHostAsync`, then the method doesn't need to async, since it no longer has a need to switch to the UI thread (use the non-async service provider). Since async methods are implemented by creating an async state machine on the head, this causes memory allocations, so we can remove the memory allocation by not using async methods when they're not needed.

Reply:
I removed `GetInfoBarHostAsync` and moved the code to this function, I wasn't able to remove the async because `uiShell` requires an async method to get the service and since its only used here I think is better to have that performance improvement.